src/main/java/com/minecolonies/colony/buildings/BuildingFarmer.java

```diff
@@ -4,131 +4,260 @@
+     */
+    public Field getFieldToWorkOn()
+    {
+        Collections.shuffle(farmerFields);
+        for(Field field: farmerFields)
+        {
+            if(field.needsWork())
+            {
+                currentField = field;
+                return field;
+            }
+        }
+        return null;
     }
 
     @Override
-    public void writeToNBT(NBTTagCompound compound)
+    public String getSchematicName()
     {
-        super.writeToNBT(compound);
-
-        NBTTagCompound farmerCompound = new NBTTagCompound();
+        return FARMER;
+    }
 
-        farmerCompound.setInteger(WHEAT_TAG,wheat);
-        farmerCompound.setInteger(POTATO_TAG,potato);
-        farmerCompound.setInteger(CARROT_TAG,carrot);
-        farmerCompound.setInteger(MELON_TAG,melon);
-        farmerCompound.setInteger(PUMPKIN_TAG, pumpkin);
+    @Override
+    public int getMaxBuildingLevel()
+    {
+        return MAX_BUILDING_LEVEL;
+    }
 
-        compound.setTag(TAG_FARMER, farmerCompound);
+    @Override
+    public String getJobName()
+    {
+        return FARMER;
     }
 
-    public static class View extends AbstractBuildingWorker.View
+    /**
+     * Synchronize field list with colony.
+     * @param world the world the building is in.
+     */
+    public void synchWithColony(World world)
     {
-        public int wheat = 100,
-                potato = 0,
-                carrot = 0,
-                melon = 0,
-                pumpkin = 0;
+        if(!farmerFields.isEmpty())
+        {
+            ArrayList<Field> tempFields = new ArrayList<>(farmerFields);
 
+            for(Field field: tempFields)
+            {
+                ScarecrowTileEntity scarecrow = (ScarecrowTileEntity) world.getTileEntity(field.getID());
+                if(scarecrow == null)
+                {
+                    farmerFields.remove(field);
+                    if (currentField != null && currentField.getID() == field.getID())
+                    {
+                        currentField = null;
+                    }
+                }
+                else
+                {
+                    field.setInventoryField(scarecrow.inventoryField);
+                    if(currentField != null && currentField.getID() == field.getID())
+                    {
+                        currentField.setInventoryField(scarecrow.inventoryField);
+                    }
+                }
+            }
+        }
+    }
 
-        public View(ColonyView c, BlockPos l)
+    /**
+     * Resets the fields to need work again.
+     */
+    public void resetFields()
+    {
+        for(Field field: farmerFields)
         {
-            super(c, l);
+            field.setNeedsWork(true);
+            field.calculateSize(getColony().getWorld(), field.getLocation().down());
         }
+    }
 
-        public com.blockout.views.Window getWindow()
+    @Override
+    public void onDestroyed()
+    {
+        super.onDestroyed();
+        for(Field field: farmerFields)
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'field' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)